### PR TITLE
Realtek: add CRC offloading and IGMP snooping to ethernet driver

### PR DESCRIPTION
**Pull request by ghost**
This adds support for CRC offloading and IGMP snooping to the Realtek Ethernet driver
on RTL8380/RTL8390 and RTL9300 SoCs.

Tested on all 3 architectures.

### CODE REVIEW

**Comment by bkobl** [comment]
Because of the comments received I added the possibility to make the CRC verification on RX configurable, the TX CRC verification is not configurable in the SoCs.

File: target/linux/realtek/files-5.4/drivers/net/ethernet/rtl838x_eth.c [hunk 1]
```diff
@@ -1837,6 +1862,40 @@ static int rtl838x_mdio_remove(struct rtl838x_eth_priv *priv)
...
+
+static int rtl83xx_set_features(struct net_device *dev, netdev_features_t features)
+{
+	struct rtl838x_eth_priv *priv = netdev_priv(dev);
+
+	if ((features ^ dev->features) & NETIF_F_RXCSUM) {
+		if (!(features & NETIF_F_RXCSUM))
+			sw_w32_mask(BIT(3), 0, priv->r->mac_port_ctrl(priv->cpu_port));
+		else
+			sw_w32_mask(0, BIT(4), priv->r->mac_port_ctrl(priv->cpu_port));
```
Comment (svanheule) [inline target/linux/realtek/files-5.4/drivers/net/ethernet/rtl838x_eth.c:1881]: Shouldn't this be BIT(3) too?

Comment (bkobl) [reply target/linux/realtek/files-5.4/drivers/net/ethernet/rtl838x_eth.c]: You are absolutely right on both accounts. Fixed.

**Comment by Borromini** [comment]
I'm seeing this, not sure if this is related to the patch set or to the kernel bump (I'm on 5.4.116 now). It sure looks bufer related though.

This is on 21.02 HEAD, so it might not be really useful, but from what I can tell master hasn't really seen any realtek code bumps since 21.02 got branched?

```
Sun May  2 21:07:06 2021 kern.info kernel: [ 7126.725932] RTL8380 Link change: status: 1, ports 200
Sun May  2 21:07:06 2021 kern.info kernel: [ 7126.732039] rtl83xx-switch switch@bb000000 lan2: Link is Down
Sun May  2 21:07:06 2021 kern.info kernel: [ 7126.738672] switch: port 2(lan2) entered disabled state
Sun May  2 21:07:06 2021 daemon.notice netifd: Network device 'lan2' link is down
Sun May  2 21:07:06 2021 kern.info kernel: [ 7126.747831] rtl83xx-switch switch@bb000000 lan2: Link is Up - 100Mbps/Full - flow control off
Sun May  2 21:07:06 2021 kern.info kernel: [ 7126.757634] switch: port 2(lan2) entered blocking state
Sun May  2 21:07:06 2021 kern.info kernel: [ 7126.763611] switch: port 2(lan2) entered forwarding state
Sun May  2 21:07:06 2021 daemon.notice netifd: Network device 'lan2' link is up
Sun May  2 21:07:06 2021 kern.info kernel: [ 7126.886142] rtl83xx-switch switch@bb000000 lan2: Link is Down
Sun May  2 21:07:06 2021 kern.info kernel: [ 7126.892765] switch: port 2(lan2) entered disabled state
Sun May  2 21:07:06 2021 daemon.notice netifd: Network device 'lan2' link is down
Sun May  2 21:07:14 2021 kern.info kernel: [ 7134.135757] RTL8380 Link change: status: 1, ports 200
Sun May  2 21:07:15 2021 kern.info kernel: [ 7135.079605] rtl83xx-switch switch@bb000000 lan2: Link is Up - 1Gbps/Full - flow control rx/tx
Sun May  2 21:07:15 2021 kern.info kernel: [ 7135.089371] switch: port 2(lan2) entered blocking state
Sun May  2 21:07:15 2021 kern.info kernel: [ 7135.095356] switch: port 2(lan2) entered forwarding state
Sun May  2 21:07:15 2021 daemon.notice netifd: Network device 'lan2' link is up
Sun May  2 21:19:48 2021 kern.info kernel: [ 7888.451509] RX buffer overrun: status 202, mask: ffdff
Sun May  2 21:20:02 2021 kern.info kernel: [ 7902.751861] RX buffer overrun: status 202, mask: ffdff
Sun May  2 21:20:03 2021 kern.info kernel: [ 7903.839621] RX buffer overrun: status 202, mask: ffdff
Sun May  2 21:20:03 2021 kern.info kernel: [ 7903.845791] RX buffer overrun: status 202, mask: ffdff
Sun May  2 21:20:03 2021 kern.info kernel: [ 7903.851735] RX buffer overrun: status 202, mask: ffdff
Sun May  2 21:20:39 2021 kern.info kernel: [ 7939.242196] RX buffer overrun: status 202, mask: ffdff
Sun May  2 21:29:48 2021 kern.info kernel: [ 8488.765135] RX buffer overrun: status 202, mask: ffdff
Sun May  2 21:29:59 2021 kern.info kernel: [ 8499.254894] RX buffer overrun: status 302, mask: ffcff
Sun May  2 21:30:00 2021 kern.info kernel: [ 8500.249148] RX buffer overrun: status 202, mask: ffdff
Sun May  2 21:30:33 2021 kern.info kernel: [ 8533.837603] RX buffer overrun: status 202, mask: ffdff
Sun May  2 21:30:33 2021 kern.info kernel: [ 8533.843766] RX buffer overrun: status 202, mask: ffcff
Sun May  2 21:30:33 2021 kern.info kernel: [ 8533.849722] RX buffer overrun: status 202, mask: ffcff
Sun May  2 21:31:14 2021 kern.info kernel: [ 8574.758923] RX buffer overrun: status 302, mask: ffcff
Sun May  2 22:09:43 2021 kern.info kernel: [10883.396083] RX buffer overrun: status 202, mask: ffdff
Sun May  2 22:09:56 2021 kern.info kernel: [10896.894751] RX buffer overrun: status 202, mask: ffdff
Sun May  2 22:11:06 2021 kern.info kernel: [10966.402423] RX buffer overrun: status 202, mask: ffdff
Sun May  2 22:19:52 2021 kern.info kernel: [11492.408196] RX buffer overrun: status 302, mask: ffcff
Sun May  2 22:21:03 2021 kern.info kernel: [11563.398458] RX buffer overrun: status 202, mask: ffdff
Sun May  2 22:26:39 2021 kern.info kernel: [11899.818395] RTL8380 Link change: status: 1, ports 200
Sun May  2 22:26:39 2021 kern.info kernel: [11899.824500] rtl83xx-switch switch@bb000000 lan2: Link is Down
Sun May  2 22:26:39 2021 kern.info kernel: [11899.831127] switch: port 2(lan2) entered disabled state
Sun May  2 22:26:39 2021 daemon.notice netifd: Network device 'lan2' link is down
Sun May  2 22:26:39 2021 kern.info kernel: [11899.840288] rtl83xx-switch switch@bb000000 lan2: Link is Up - 1Gbps/Full - flow control rx/tx
Sun May  2 22:26:39 2021 kern.info kernel: [11899.850332] switch: port 2(lan2) entered blocking state
Sun May  2 22:26:39 2021 kern.info kernel: [11899.856319] switch: port 2(lan2) entered forwarding state
Sun May  2 22:26:39 2021 daemon.notice netifd: Network device 'lan2' link is up
Sun May  2 22:26:40 2021 kern.info kernel: [11900.774138] rtl83xx-switch switch@bb000000 lan2: Link is Down
Sun May  2 22:26:40 2021 kern.info kernel: [11900.780758] switch: port 2(lan2) entered disabled state
Sun May  2 22:26:40 2021 daemon.notice netifd: Network device 'lan2' link is down
Sun May  2 22:26:42 2021 kern.info kernel: [11902.433654] RTL8380 Link change: status: 1, ports 200
Sun May  2 22:26:42 2021 kern.info kernel: [11902.823534] rtl83xx-switch switch@bb000000 lan2: Link is Up - 10Mbps/Full - flow control rx/tx
Sun May  2 22:26:42 2021 kern.info kernel: [11902.833339] switch: port 2(lan2) entered blocking state
Sun May  2 22:26:42 2021 kern.info kernel: [11902.839318] switch: port 2(lan2) entered forwarding state
Sun May  2 22:26:42 2021 daemon.notice netifd: Network device 'lan2' link is up
Sun May  2 22:28:21 2021 authpriv.info dropbear[2938]: Child connection from 10.0.0.5:58678
Sun May  2 22:28:21 2021 authpriv.notice dropbear[2938]: Pubkey auth succeeded for 'root' with key sha1!! c1:e0:fb:73:e0:1e:f8:c1:52:65:5d:69:74:79:1c:f3:20:b1:54:b1 from 10.0.0.5:58678
Sun May  2 22:29:41 2021 kern.info kernel: [12081.916642] RX buffer overrun: status 202, mask: ffdff
Sun May  2 22:29:53 2021 kern.info kernel: [12093.914075] RX buffer overrun: status 202, mask: ffdff
Sun May  2 22:29:54 2021 kern.info kernel: [12094.913628] RX buffer overrun: status 202, mask: ffdff
Sun May  2 22:30:28 2021 kern.info kernel: [12128.914243] RX buffer overrun: status 302, mask: ffcff
Sun May  2 22:30:28 2021 kern.warn kernel: [12128.920270] rtl838x-eth bb00a300.ethernet eth0: Ring contention: r: 1, last a70001dc, cur a70001e0
Sun May  2 22:30:28 2021 kern.warn kernel: [12128.930399] rtl838x-eth bb00a300.ethernet eth0: Ring contention: r: 1, last a70001dc, cur a70001e0
Sun May  2 22:30:28 2021 kern.warn kernel: [12128.941311] rtl838x-eth bb00a300.ethernet eth0: Ring contention: r: 1, last a70001dc, cur a70001e0
Sun May  2 22:30:29 2021 kern.warn kernel: [12128.955101] rtl838x-eth bb00a300.ethernet eth0: Ring contention: r: 1, last a70001dc, cur a70001e0
Sun May  2 22:30:29 2021 kern.info kernel: [12129.903400] RX buffer overrun: status 202, mask: ffdff
Sun May  2 22:30:29 2021 kern.info kernel: [12129.909283] RX buffer overrun: status 2, mask: ffdff
Sun May  2 22:30:29 2021 kern.info kernel: [12129.915158] RX buffer overrun: status 202, mask: ffdff
Sun May  2 22:30:29 2021 kern.info kernel: [12129.921087] RX buffer overrun: status 2, mask: ffdff
Sun May  2 22:30:29 2021 kern.info kernel: [12129.926717] RX buffer overrun: status 2, mask: ffdff
```

**Comment by bkobl** [comment]
These debug messages should appear also on the original kernel if you start to transfer a lot of data to the switch or back from it. This is just some output that shows the switch is running into the limit of the CPU not being able to move data received from the receive buffer. The errors should have also appeared without the patch: https://forum.openwrt.org/t/support-for-rtl838x-based-managed-switches/57875/477?u=kobi
Did you test with e.g. iperf?

**Comment by Borromini** [comment]
No, this was during a few movie streams from the home server to the HTPC. Good to know it's normal behaviour, thanks.

**Comment by bkobl** [comment]
The code in this PR is also in https://github.com/openwrt/openwrt/pull/4131 , this can be closed.

**Comment by adschm** [comment]
Closed as requested.

**Pull request closed without merging**